packages/jaeger-ui/src/components/common/FilteredList/highlightMatches.tsx

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as React from 'react';
+import memoize from 'lru-memoize';
+
+enum EHighlightCadence {
```

Comment:
I dug into a bit more, but seems like simple RegExp matching doesn't work as `EHighlightCadence` would suggest. Without the `g` flag (which matchers don't use), only the first match is returned so my `toHighlights` function will handle it. With the `g` flag, the non matches will be dropped and capture groups will be ignored, and that doesn't directly fit this use case and would require more work to use.

Reply:
> Defining this `EHighlightCadence` and specifying it in `MatchHiglighter.matchers` allows the clever `if (i % 2 !== cadence)`, but it seems obfuscated away from its true purpose.

I don't think I made the true purpose sufficiently clear as your alternate definition of `toHighlights` won't work.

If the use of the cadence seems flashy, then I have not made the necessity of it clear, at all.

>It seems as though the distinction isn't odd versus even. Rather it's "if the RegExp contains ^".

Odd vs even covers all of the scenarios. Whether or not the regex contains `^` or not is not enough, hence the alternate `toHighlights` doesn't work.

For the search term `eg`, five regexes are created (as noted in the code comments):

1. `/^(eg)(.*)/i`
2. `/(.*?)(\b[eE][gG]|Eg)(.*)/`
3. `/(.*?)(\be|E)(.*?)(\bg|G)(.*)/`
4. `/(.*?)(eg)(.*)/i`
5. `/(.*?)(e)(.*?)(g)(.*)/i`

The alternate function you provided would work for 1, 2 and 4.

3 and 5 have `2 * n + 1` groups where `n` is the number of characters in the search term.

```
/(.*?)(e)(.*?)(g)(.*)/i
   ^   ^   ^   ^  ^
   Five capturing groups
```

Since regexes 3 and 5 are of unknown length, you have to use the `i % 2`, in a loop, any way you spin it.

For all of the groups, text that should be highlighted is interleaved with text that shouldn't be highlighted. The only distinction, in this respect, is whether or not text to highlight starts at `match[1]` or `match[2]`.

Please let me know if you have a suggestion that is less clever, less flashy and will work for all five scenarios.

Reply:
>Again, not sure if this should be handled in the first pass, but it may make sense to capture this lack of multiple match support in a comment.

<img width="607" alt="Screen Shot 2019-07-02 at 10 30 31 PM" src="https://user-images.githubusercontent.com/2304337/60559340-19425980-9d1a-11e9-8344-70b551ec471f.png">

In the image above, the second item has many more potential matches than the first item, but it is ranked second. I don't believe the ranking takes into consideration multiple matches, so I don't think the highlighting should either.